riscv/processor.cc

```diff
@@ -1523,25 +1513,47 @@ reg_t processor_t::get_csr(int which, insn_t insn, bool write, bool peek)
       ret((VU.vxsat << VCSR_VXSAT_SHIFT) | (VU.vxrm << VCSR_VXRM_SHIFT));
     case CSR_INSTRET:
     case CSR_CYCLE:
-      if (!ctr_ok)
-        goto throw_illegal;
-      if (!ctr_v_ok)
-        goto throw_virtual;
-      ret(state.minstret);
+    case CSR_HPMCOUNTER3 ... CSR_HPMCOUNTER31:
+      if (!ctr_ok(which)) {
+        if (state.v)
+          goto throw_virtual;
+        else
+          goto throw_illegal;
+      }
+      if (which == CSR_INSTRET || which == CSR_CYCLE)
+        ret(state.minstret);
+      else
+        ret(0);
     case CSR_MINSTRET:
     case CSR_MCYCLE:
-      ret(state.minstret);
+    case CSR_MHPMCOUNTER3 ... CSR_MHPMCOUNTER31:
+    case CSR_MHPMEVENT3 ... CSR_MHPMEVENT31:
+      if (which == CSR_MINSTRET || which == CSR_MCYCLE)
+        ret(state.minstret);
+      else
+        ret(0);
     case CSR_INSTRETH:
     case CSR_CYCLEH:
-      if (!ctr_ok || xlen != 32)
-        goto throw_illegal;
-      if (!ctr_v_ok)
-        goto throw_virtual;
-      ret(state.minstret >> 32);
+    case CSR_HPMCOUNTER3H ... CSR_HPMCOUNTER31H:
+      if (!ctr_ok(which) || xlen != 32) {
```

Comment:
This still appears to throw virtual instead of illegal if `xlen==32 && state.v`. Is that right?

Reply:
This case is not clearly stated in H-extension spec.

I think it's better to generate virtual instruction trap when RV64 guest access high-half CSRs so that hypervisor directly gets trap without involving M-mode firmware. If we generate illegal instruction trap then M-mode firmware will take the trap and then redirect it to Hypervisor.

@aswaterman what do you think?

Regards,
Anup

Reply:
The spec does address this case explicitly: "When V=1 and XLEN>32, an attempt to access a high-half supervisor-level CSR, high-half hypervisor CSR, high-half VS CSR, or high-half unprivileged CSR always raises an illegal instruction exception."

It doesn't make sense to send this exception to the hypervisor because there's nothing interesting to emulate: it's actually an illegal instruction.

Reply:
Okay, I will update this PR.

This would mean that M-mode firmware will take illegal instruction trap which will be further redirected to Hypervisor and  Hypervisor will again redirect it to Guest/VM.

Regards,
Anup